### DISCUSSION TITLE
What does "Only allowed in Babel's **programmatic options** " mean?

### DISCUSSION THREAD

**Discussion by nyngwang** [babel/babel#14405] [Q&A] [answered]
As title. I found the term used in the documentation very unclear. For example, under [options#primary-options](https://babeljs.io/docs/en/options#primary-options):

> These options are only allowed as part of Babel's **programmatic options**,  
so they are primarily for use by tools that wrap around Babel, or people calling babel.transform directly.  
Users of Babel's integrations, like babel-loader or [@babel/register](https://babeljs.io/docs/en/babel-core#options) are unlikely to use these.

I cannot find anywhere the term "programmatic options" is defined. The only related question I found on StackOverflow is this one: https://stackoverflow.com/q/52957510/5290519, where a comment said that "programmatic options" should include the option of `babel-loader` from `webpack.config.js`, and this contradicts the statement above:

> [...], like **babel-loader** or [...] are unlikely to use these.

## Update

In the first paragraph of the same page [options#primary-options](https://babeljs.io/docs/en/options#primary-options):

> Options can be passed to Babel in a variety of ways. **"When passed directly to Babel"**, you can just pass the options object.

The part I put emphasis on is very unclear too(at least for me). People who are new to Babel probably won't know about which ways would fit the adverb "directly". For example, *in my definition*, an argument passed to the `@babel/cli` can fit the term "directly". So, IMOO, this part of the documentation should be updated.

**Comment by nicolo-ribaudo** [comment, marked as answer]
It's when you specify them directly to Babel (`babel.transformSync(code, programmaticOptions)` or to the Babel integration you are using (e.g. `babel-loader`, which can pass them to its internal `babel.transform` call). In other words, not in presets or config files.

"Users are unlikely to use these" is referring to that specific "Primary options" section, not to all programmatic options.

**Reply by nyngwang** [reply]
Do you accept people's PRs for updating the documentation? I also found an error(i.e. one link is obsolete) on the other page.

**Reply by nicolo-ribaudo** [reply]
Yes! They are hosted in the `babel/website` repository.

**Reply by nyngwang** [reply]
@nicolo-ribaudo Could you also explain [this line](https://babeljs.io/docs/en/options):

> When Babel is **used via a wrapper**, it may also be necessary, or at least more useful, to pass the options via [configuration files](https://babeljs.io/docs/en/config-files).

So in which cases it will be necessary, and in which ones be not necessary but useful, respectively? Thanks in advance.

**Reply by nicolo-ribaudo** [reply]
A custom library (wrapper) that uses Babel may not allow you to pass some options to Babel. If you use a config file, Babel will read it regardless of what the library passes through or not.

It's not necessary when using the main Babel integrations (`babel-loader`, `@rollup/plugin-babel`, `@eslint/babel-parser`) because they pass through all the options, but it still allows you to more clearly separate "Babel config" from (for example) "Webpack config".

**Reply by nyngwang** [reply]
@nicolo-ribaudo Can I conclude like this: Under [configuration - #print-effective-configs](https://babeljs.io/docs/en/configuration#print-effective-configs) the first three printed-`config`s are from the two config files `babel.config.json`(project-level),`.babelrc`(file-level), and the last item is `programmatic options from ...`, which confirms your "[...] In other words, not in presets or config files.". This also means, implicitly, that a custom library cannot stop you from passing info. into Babel using configs, since Babel always read them.

**Reply by nicolo-ribaudo** [reply]
Uh well, a custom library could pass `configFile: false, babelrc: false` to Babel's programmatic options to prevent it from reading the config!
